src/vm/checker/typecheck/mod.rs

```diff
@@ -0,0 +1,404 @@
+    if !expected.admits_type(to_check) {
+        Err(CheckError::new(CheckErrors::TypeError(expected, to_check.clone())))
+    } else {
+        Ok(())
+    }
+}
+
+impl <'a> TypeChecker <'a> {
+    pub fn new(db: &'a AnalysisDatabase) -> TypeChecker {
+        TypeChecker {
+            db: db,
+            contract_context: ContractContext::new(),
+            type_map: TypeMap::new()
+        }
+    }
+
+    fn type_check_all(&mut self, args: &[SymbolicExpression], context: &TypingContext) -> CheckResult<Vec<TypeSignature>> {
+        let mut result = Vec::new();
+        for arg in args.iter() {
+            // don't use map here, since type_check has side-effects.
+            result.push(self.type_check(arg, context)?)
+        }
+        Ok(result)
+    }
+
+    fn type_check_function_type(&mut self, func_type: &FunctionType,
+                                args: &[SymbolicExpression], context: &TypingContext) -> TypeResult {
+        let typed_args = self.type_check_all(args, context)?;
+        func_type.check_args(&typed_args)?;
+        Ok(func_type.return_type().clone())
+    }
+
+    fn type_check_list_pairs<'b> (&mut self, bindings: &[SymbolicExpression],
+                                  context: &'b TypingContext) -> CheckResult<TypingContext<'b>> {
+        let mut out_context = context.extend()?;
+        for binding in bindings.iter() {
+            let binding_exps = binding.match_list()
+                .ok_or(CheckError::new(CheckErrors::BadSyntaxBinding))?;
+            
+            if binding_exps.len() != 2 {
+                return Err(CheckError::new(CheckErrors::BadSyntaxBinding))
+            }
+
+            let var_name = binding_exps[0].match_atom()
+                .ok_or(CheckError::new(CheckErrors::BadSyntaxBinding))?;
+
+            self.type_map.set_type(&binding_exps[0], no_type())?;
+            let typed_result = self.type_check(&binding_exps[1], context)?;
+            out_context.variable_types.insert(var_name.clone(),
+                                              typed_result);
+        }
+
+        Ok(out_context)
+    }
+
+    fn get_function_type(&self, function_name: &str) -> Option<FunctionType> {
+        if let Some(function_type) = self.contract_context.get_function_type(function_name) {
+            Some(function_type.clone())
+        } else {
+            None
+        }
+    }
+
+    fn type_check_define_function(&mut self, function_expression: &[SymbolicExpression],
+                                  context: &TypingContext) -> CheckResult<(String, FunctionType)> {
+        if function_expression.len() != 3 {
+            return Err(CheckError::new(CheckErrors::IncorrectArgumentCount(2, function_expression.len() - 1)))
+        }
+
+        self.type_map.set_type(&function_expression[0], no_type())?;
+        self.type_map.set_type(&function_expression[1], no_type())?;
+        // should we set the type of the subexpressions of the signature to no-type as well?
```

Comment:
Can you give an example?

Reply:
Right, so let's say you're defining a function:

```
(define (foo (a int) (b int)) 
   (+ a b))
```

The AST looks like:
```
           |
  -----------------        ....
/           |
define    / | \
       foo  *  *
           / \ | \
         /   / |  \
       a  int  b  int
              
```

Those two lines set the two "argument expressions" to `define` to no-type: that's the signature `(foo (a int) (b int))` and then the function definition itself.

Now, it's never actually terribly important to figure out the types of those arguments, because the type checker will never actually want them -- it doesn't try to evaluate them directly, rather, it looks for the FunctionType object. Also, those arguments don't really have types anyways. Now, the question is just whether for sake of ensuring some property like

Every expression in the AST gets assigned a type or NoType,

should we iterate over subexpressions of untyped things (in the case of above, something like `(a int)` is a subexpression of the argument to `define`) and assign each of them all NoType ?